build-tools/packages/build-cli/src/library/commands/generateSourceEntrypoints.ts

```diff
@@ -0,0 +1,336 @@
+		}),
+		srcDir: Flags.string({
+			description: "Directory to emit source entrypoint declaration files.",
+			default: optionDefaults.srcDir,
+			exists: true,
+		}),
+		...BaseCommand.flags,
+	};
+
+	public async run(): Promise<void> {
+		const { mainEntrypoint } = this.flags;
+
+		const packageJson = await readPackageJson();
+
+		const tsConfig = await readTsConfig();
+
+		const { mapQueryPathToApiTagLevel, mapApiTagLevelToOutput } = getOutputConfiguration(
+			this.flags,
+			packageJson,
+			tsConfig,
+			this.logger,
+		);
+
+		if (mapApiTagLevelToOutput.size === 0) {
+			throw new Error(
+				`There are no package exports matching requested output entrypoints:\n\t${[
+					...mapQueryPathToApiTagLevel.keys(),
+				].join("\n\t")}`,
+			);
+		}
+
+		const promises: Promise<void>[] = [];
+
+		// In the past @alpha APIs could be mapped to /legacy via --outFileAlpha.
+		// When @alpha is mapped to /legacy, @beta should not be included in
+		// @alpha aka /legacy entrypoint.
+		const separateBetaFromAlpha = this.flags.outFileAlpha !== ApiLevel.alpha;
+		promises.push(
+			generateSourceEntrypoints(
+				mainEntrypoint,
+				mapApiTagLevelToOutput,
+				this.logger,
+				separateBetaFromAlpha,
+			),
+		);
+
+		// All of the output actions (deletes of stale files or writing of new/updated files)
+		// are all independent and can be done in parallel.
+		await Promise.all(promises);
+	}
+}
+
+async function readPackageJson(): Promise<PackageJson> {
+	const packageJson = await fs.readFile("./package.json", { encoding: "utf8" });
+	return JSON.parse(packageJson) as PackageJson;
+}
+
+async function readTsConfig(): Promise<TsConfigJson> {
+	const tsConfigContent = await fs.readFile("./tsconfig.json", { encoding: "utf8" });
+	// Trim content to avoid unexpected whitespace issues
+	const trimmedContent = tsConfigContent.trim();
+
+	// Remove trailing commas before parsing
+	const sanitizedContent = trimmedContent.replace(/,\s*([\]}])/g, "$1");
+
+	// Parse and validate JSON content
+	return JSON.parse(sanitizedContent) as TsConfigJson;
+}
+
+function getOutputConfiguration(
+	flags: Readonly<Record<keyof typeof optionDefaults, string>>,
+	packageJson: PackageJson,
+	tsconfig: TsConfigJson,
+	logger?: CommandLogger,
+): {
+	mapQueryPathToApiTagLevel: Map<string | RegExp, ApiTag | undefined>;
+	mapApiTagLevelToOutput: Map<ApiTag, ExportData>;
+} {
+	const { outFileSuffix, outFileAlpha, outFileBeta, outFileLegacy, outFilePublic, srcDir } =
+		flags;
+
+	const mapQuerySrcPathToApiTagLevel: Map<string | RegExp, ApiTag | undefined> = new Map([
+		[`${srcDir}${outFileAlpha}${outFileSuffix}`, ApiTag.alpha],
+		[`${srcDir}${outFileBeta}${outFileSuffix}`, ApiTag.beta],
+		[`${srcDir}${outFilePublic}${outFileSuffix}`, ApiTag.public],
+	]);
+
+	// In the past @alpha APIs could be mapped to /legacy via --outFileAlpha.
+	// If @alpha is not mapped to same as @legacy, then @legacy can be mapped.
+	if (outFileAlpha !== outFileLegacy) {
```

Comment:
New command - doesn't need back compat for old modes. legacy can just be in the map initialization above.